lua/lean/html.lua

```diff
@@ -557,93 +558,99 @@ function Div:buf_hover(force_update_highlight)
+      self.tooltip.disable_update = false
       -- workaround for neovim/neovim#13403, as it seems this wasn't entirely resolved by neovim/neovim#14770
       vim.api.nvim_command("redraw")
-      bufdata.tooltip._bufdata.last_win_options = vim.deepcopy(win_options)
-    elseif not vim.deep_equal(win_options, bufdata.tooltip._bufdata.last_win_options) then
-      vim.api.nvim_win_set_config(bufdata.tooltip._bufdata.last_win, win_options)
+      self.tooltip.last_win_options = vim.deepcopy(win_options)
+    elseif not vim.deep_equal(win_options, self.tooltip.last_win_options) then
+      vim.api.nvim_win_set_config(self.tooltip.last_win, win_options)
       vim.api.nvim_command("redraw")
-      bufdata.tooltip._bufdata.last_win_options = vim.deepcopy(win_options)
+      self.tooltip.last_win_options = vim.deepcopy(win_options)
     end
 
-    bufdata.tooltip:buf_render()
+    self.tooltip:buf_render()
   end
 
   if hover_div_path and hover_div then
-    local a = self:pos_from_path(hover_div_path)
-    local start_pos = raw_pos_to_pos(a, bufdata.lines)
-    local end_pos = raw_pos_to_pos(a + hover_div._size, bufdata.lines)
-    bufdata.hover_range = {start_pos, end_pos}
+    local a = self.div:pos_from_path(hover_div_path)
+    local start_pos = raw_pos_to_pos(a, self.lines)
+    local end_pos = raw_pos_to_pos(a + hover_div._size, self.lines)
+    self.hover_range = {start_pos, end_pos}
   else
-    bufdata.hover_range = nil
+    self.hover_range = nil
   end
 
-  if force_update_highlight or not vim.deep_equal(old_hover_range, bufdata.hover_range) then
-    vim.api.nvim_buf_clear_namespace(bufdata.buf, hl_ns, 0, -1)
+  if force_update_highlight or not vim.deep_equal(old_hover_range, self.hover_range) then
+    vim.api.nvim_buf_clear_namespace(self.buf, hl_ns, 0, -1)
     local hlgroup = "htmlDivHighlight"
-    if bufdata.hover_range then
-      vim.highlight.range(bufdata.buf, hl_ns, hlgroup, bufdata.hover_range[1], bufdata.hover_range[2])
+    if self.hover_range then
+      vim.highlight.range(self.buf, hl_ns, hlgroup, self.hover_range[1], self.hover_range[2])
     end
   end
 end
 
-function Div:buf_event(event, path, ...)
+function BufDiv:buf_event(event, path, ...)
   local args = {...}
-  local bufdata = self._bufdata
 
-  path = path or bufdata.path
+  path = path or self.path
 
-  if not self:event(path, event, self:buf_make_event_context(), unpack(args)) and bufdata.parent then
+  if not self.div:event(path, event, self:buf_make_event_context(), unpack(args)) and self.parent then
     -- bubble up to parent
-    return bufdata.parent:buf_event(event, bufdata.parent_path, ...)
+    return self.parent:buf_event(event, self.parent_path, ...)
   end
 end
 
-function Div:buf_enter_win()
+---@return DivEventContext
+function BufDiv:buf_make_event_context()
+  return {
+    rerender = function() self:buf_render() end,
+    rehover = function() self:buf_hover() end,
+    root = self,
```

Comment:
BTW, this is no longer type correct now.

Reply:
Good catch. But I'm thinking we just do away with this function entirely and find a way to get the render function at the component level.